tests/mesh/test-cartesian-grid.cpp

```diff
@@ -0,0 +1,144 @@
+/*!
+ * @file Tests for the frame structures and the cartesian grids
+ * @author Melchior Schuh-Senlis
+ */
+
+using namespace RINGMesh;
+
+void test_frames_and_cartesian_grid()
+{
+    vec3 origin{ 100, 200, -4 };
+    vec3 x{ 1, 2, -1 };
+    vec3 y{ 2, 1, 4 };
+    vec3 z{ 3, -2, -1 };
+
+    Frame3D frame{ x, y, z };
+    ReferenceFrame3D reference_frame{ origin, frame };
+
+    vec3 test_coordinates{ 20, 10, 30 };
+    vec3 frame_test_coordinates =
+        ReferenceFrameManipulator3D::coords_from_global_to_frame(
+            reference_frame, test_coordinates );
+    vec3 operation_test_coordinates =
+        ReferenceFrameManipulator3D::coords_from_frame_to_global(
+            reference_frame, frame_test_coordinates );
+    if( !inexact_equal( test_coordinates, operation_test_coordinates, 1e-7 ) )
+    {
+        throw RINGMeshException(
+            "TEST", "Error in coordinate reference change" );
+    }
+    if( !ReferenceFrameManipulator3D::is_frame_orthogonal( reference_frame ) )
+    {
+        throw RINGMeshException(
+            "TEST", "Error in checking the orthogonality of a frame" );
+    }
+
+    ReferenceFrame3D inverse_frame =
+        ReferenceFrameManipulator3D::reference_frame_from_global_to_local(
+            reference_frame );
+    if( !inexact_equal( inverse_frame,
+            ReferenceFrameManipulator3D::
+                orthogonal_reference_frame_from_global_to_local(
+                    reference_frame ),
+            1e-7 ) )
+    {
+        throw RINGMeshException(
+            "TEST", "Error in orthogonal reference frame change" );
+    }
+
+    ivec3 grid_dimensions{ 10, 8, 9 };
+    CartesianGrid3D cartesiangrid{ grid_dimensions, reference_frame };
+    if( cartesiangrid.cell_volume() != 42. )
+    {
+        throw RINGMeshException( "TEST", "Error in cell volume" );
+    }
+    if( cartesiangrid.cell_offset_from_global_point( vec3{ 103, 200.5, -3 } )
+        != 0 )
+    {
+        throw RINGMeshException( "TEST", "Error in calculating the offset 1" );
+    }
+    if( cartesiangrid.cell_offset_from_global_point( vec3{ 107, 201.5, -1 } )
+        != 91 )
+    {
+        throw RINGMeshException( "TEST", "Error in calculating the offset 2" );
+    }
+    if( cartesiangrid.cell_offset_from_global_point( vec3{ 1, 1, 1 } ) != -1 )
+    {
+        throw RINGMeshException( "TEST", "Error in calculating the offset 3" );
+    }
+    if( cartesiangrid.cell_offset( cartesiangrid.local_from_offset( 51 ) )
+        != 51 )
+    {
+        throw RINGMeshException( "TEST", "Error in calculating the inverse offset 1" );
+    }
+    if( cartesiangrid.local_from_offset( cartesiangrid.cell_offset( sivec3{ 5, 2, 8 } ) )
+        != sivec3{ 5, 2, 8 } )
+    {
+        throw RINGMeshException( "TEST", "Error in calculating the inverse offset 2" );
+    }
+}
+
+int main()
+{
+    using namespace RINGMesh;
+
+    try
+    {
+        Logger::out( "TEST", "Frames and Cartesian Grid" );
+
+        test_frames_and_cartesian_grid();
```

Comment:
You can split your tests into two functions: test_frames() and test_cartesian_grids();